aiohttp/web.py

```diff
@@ -647,6 +662,90 @@ def write_eof(self):
         yield from super().write_eof()
 
 
+class WebSocketResponse(StreamResponse):
+
+    def __init__(self, *protocols):
+        super().__init__(status=101)
+        self._protocols = protocols
+        self._protocol = None
+        self._writer = None
+        self._reader = None
+
+    def start(self, request):
+        # make pre-check to don't hide it by do_handshake() exceptions
+        resp_impl = self._start_pre_check(request)
+        if resp_impl is not None:
+            return resp_impl
+
+        status, headers, parser, writer, protocol = do_handshake(
+            request.method, request.headers, request.transport,
+            self._protocols)
+
+        if self.status != status:
+            self.set_status(status)
+        for k, v in headers:
+            self.headers[k] = v
+
+        resp_impl = super().start(request)
+
+        self._reader = request._reader.set_parser(parser)
+        self._writer = writer
+        self._protocol = protocol
+
+        return resp_impl
+
+    @property
+    def protocol(self):
+        return self._protocol
+
+    def ping(self):
+        if self._writer is None:
+            raise RuntimeError('Call .start() first')
+        self._writer.ping()
+
+    def send_str(self, data):
+        if self._writer is None:
+            raise RuntimeError('Call .start() first')
+        if not isinstance(data, str):
+            raise TypeError('data argument must be str (%r)', type(data))
+        self._writer.send(data, binary=False)
+
+    def send_bytes(self, data):
+        if self._writer is None:
+            raise RuntimeError('Call .start() first')
+        if not isinstance(data, (bytes, bytearray, memoryview)):
+            raise TypeError('data argument must be byte-ish (%r)',
+                            type(data))
+        self._writer.send(data, binary=True)
+
+    def close(self, *, code=1000, message=b''):
+        if self._writer is None:
+            raise RuntimeError('Call .start() first')
+        self._writer.close(code, message)
+
+    @asyncio.coroutine
+    def receive(self):
+        if self._reader is None:
+            raise RuntimeError('Call .start() first')
+        while True:
+            try:
+                msg = yield from self._reader.read()
+            except Exception as exc:
+                raise WebSocketClosed(code=None, message=str(exc)) from exc
+
+            if msg.tp == MSG_PING:
+                self._writer.pong()
+            elif msg.tp == MSG_CLOSE:
+                raise WebSocketClosed(msg.data, msg.extra)
```

Comment:
Yes, I've rereaded WebSocket spec after pushing the PR.

I agree with `closing` state but what user should do in case of `.send_str()` after `.close()`?
Checking for flag is annoying, I guess to make `.send_*`, `.ping` and `.close` coroutines (think about multiple tasks for asynchronous sending websocket notifications).
`.send_str` should return immediately if websocket  is not in closing state or wait for closing otherwise. 
Any exception accrued on waiting should be send to any waiters.
We need for timeout for closing handhsake -- if timeout is gone all waiters should get  `WebSocketClosed` exception and socket transport should be closed directly.